h2o-py/h2o/display.py

```diff
@@ -6,126 +6,662 @@
+                trunc_table += [v for v in self._table[:first]]
+                trunc_table.append(["---"] * nc)
+                trunc_table += [v for v in self._table[(nr - last):]]
+                self._display_table = trunc_table
+                self._truncated = True  # due to the --- row, we can't trust the display_table shape 
+            else:
+                self._display_table = self._table
 
+    def _str_(self):
+        table = self._display_table
+        table_str = (table.to_string() if H2OTableDisplay.is_pandas(table)
+                     else tabulate.tabulate(table,
+                                            headers=self._columns_labels or (),
+                                            **self._kwargs))
+        return format_to_multiline([self._caption, table_str]) if self._caption else table_str
+
+    def _str_html_(self):
+        table = self._display_table
+        if H2OTableDisplay.is_pandas(table):
+            html = (table.style.set_caption(self._caption)
+                               .set_table_styles([dict(selector="",
+                                                       props=[("margin-top", "1em"),
+                                                              ("margin-bottom", "1em")]),
+                                                  dict(selector="caption",
+                                                       props=[("font-size", "larger"),
+                                                              ("text-align", "left"),
+                                                              ("white-space", "nowrap")])],
+                                                 overwrite=False)
+                               .to_html())
+        
+        else:
+            html = H2OTableDisplay._html_table(table, caption=self._caption, column_labels=self._columns_labels)
+        return html+H2OTableDisplay.table_footer(self, fmt='html') if self.truncated else html
+        
     # some html table builder helper things
     @staticmethod
-    def _html_table(rows, header=None):
-        # keep table in a div for scrollability
-        table = "<div style=\"overflow:auto\"><table style=\"width:50%\">{}</table></div>"
-        table_rows = []
-        if header is not None:
-            table_rows.append(H2ODisplay._html_row(header, bold=True))
+    def _html_table(rows, caption=None, column_labels=None):
+        html = """
+<style>
+{css}
+</style>      
+<div class="h2o-container">
+  <table class="h2o-table">
+    <caption>{caption}</caption>
+    <thead>{head}</thead>
+    <tbody>{body}</tbody>
+  </table>
+</div>
+"""
+        css = """
+.h2o-container {
+  overflow-x: auto;
+}
+.h2o-table {
+  /* width: 100%; */
+  margin-top: 1em;
+  margin-bottom: 1em;
+}
+.h2o-table caption {
+  white-space: nowrap;
+  caption-side: top;
+  text-align: left;
+  /* margin-left: 1em; */
+  margin: 0;
+  font-size: larger;
+}
+.h2o-table thead {
+  white-space: nowrap; 
+  background-color: #A0A0A0;
```

Comment:
I'm not sure if we should set colors anywhere - jupyter lab has a dark mode, so does VS code, dataspell etc.

There doesn't seem to be a way to detect if user uses the dark mode or not from python - afaik the theme is just on the client side and one python kernel could be potentially running in both dark and light themed labs/notebooks.

There is a [way](https://github.com/jupyterlab/jupyterlab/pull/5232) to tell jupyter that we require light background but I don't know if it's worth the complexity and/or if it works elsewhere as well.

It's possible that this won't pose any real problem since this background makes the text legible in both themes but there are other user-created themes and of course VS Code, DataSpell, Spyder etc.

Dark mode:
<img width="253" alt="Screen Shot 2022-06-27 at 16 32 49" src="https://user-images.githubusercontent.com/61695433/175966387-19241247-0d42-4880-a367-e198d4367e2f.png">

Light mode:
<img width="282" alt="Screen Shot 2022-06-27 at 16 32 35" src="https://user-images.githubusercontent.com/61695433/175966632-5f576eaa-c80a-4ea1-9862-f01b37aeeb4f.png">

Reply:
I chose this shade of grey because (I think) it's readable in both default ligth and dark themes in Jupyter. Didn't think about other themes…
I think the tables look better this way, especially as I made them scrollable and keeping the header row fixed… I'll try to remove the background color and see if it's not confusing when scrolling.